src/cargo/core/registry.rs

```diff
@@ -718,3 +743,114 @@ fn lock(
         dep
     })
 }
+
+/// This is a helper for selecting the summary, or generating a helpful error message.
+fn summary_for_patch(
+    orig_patch: &Dependency,
+    locked: &Option<(Dependency, PackageId)>,
+    mut summaries: Vec<Summary>,
+    source: &mut dyn Source,
+) -> CargoResult<(Summary, Option<PackageId>)> {
+    if summaries.len() == 1 {
+        return Ok((summaries.pop().unwrap(), None));
+    }
+    let best_summary = |summaries: &mut Vec<Summary>| -> Summary {
+        // TODO: This could maybe honor -Zminimal-versions?
+        summaries.sort_by(|a, b| a.version().cmp(b.version()));
+        summaries.pop().unwrap()
+    };
+    if summaries.len() > 1 {
+        let summary = best_summary(&mut summaries);
+        if let Some((_dep, lock_id)) = locked {
+            // I can't think of a scenario where this might happen (locked by
+            // definition should only match at most one summary). Maybe if the
+            // source is broken?
+            return Ok((summary, Some(*lock_id)));
+        } else {
+            return Ok((summary, None));
+        }
+    }
+    assert!(summaries.is_empty());
+    // No summaries found, try to help the user figure out what is wrong.
+    if let Some((locked_patch, locked_id)) = locked {
+        // Since the locked patch did not match anything, try the unlocked one.
+        let mut orig_matches = match source.query_vec(orig_patch) {
+            Ok(summaries) => summaries,
+            Err(e) => {
+                log::warn!(
+                    "could not determine unlocked summaries for dep {:?}: {:?}",
+                    orig_patch,
+                    e
+                );
+                Vec::new()
+            }
+        };
+        if orig_matches.is_empty() {
+            // This should be relatively unusual. For example, a patch of
+            // {version="0.1.2", ...} and the patch location no longer contains a
+            // version that matches "0.1.2". It is unusual to explicitly write a
+            // version in the patch.
+            anyhow::bail!(
+                "The patch is locked to {} in Cargo.lock, but the version in the \
+                patch location does not match any packages in the patch location.\n\
+                Make sure the patch points to the correct version.",
+                locked_patch.version_req(),
+            );
+        }
+        let summary = best_summary(&mut orig_matches);
```

Comment:
It might be best to restructure this so it's maybe recursive or otherwise shares logic with the above. I'm a bit wary of having two instances of "pick the summary out of this list of summaries" which differ slightly